tremor-connectors-azure/src/monitor/ingest/writer/sink.rs

```diff
@@ -0,0 +1,296 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// #![cfg_attr(coverage, no_coverage)] // NOTE We let this hang wet - no Azure env for CI / coverage
+
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+use log::error;
+use tremor_config::NameWithConfig;
+
+use crate::monitor::ingest::writer::RequestBuilder;
+use crate::rest::extract_request_meta;
+use crate::rest::extract_response_meta;
+use crate::rest::RequestId;
+
+use super::Config;
+use tokio::sync::mpsc::Sender;
+use tokio::time::timeout;
+use tremor_common::time::nanotime;
+use tremor_connectors::{
+    sink::{concurrency_cap::ConcurrencyCap, prelude::*},
+    source::SourceReply,
+};
+use tremor_script::prelude::MutableObject;
+use tremor_script::prelude::{literal, ValueObjectAccess};
+use tremor_script::EventOriginUri;
+
+pub(crate) struct AmiSink {
+    request_counter: u64,
+    client: Option<Arc<dyn azure_core::HttpClient>>,
+    config: Config,
+    response_tx: Sender<SourceReply>,
+    reply_tx: ReplySender,
+    concurrency_cap: ConcurrencyCap,
+    origin_uri: EventOriginUri,
+    // we should only send responses down the channel when we know there is a source consuming them
+    // otherwise the channel would fill up and we'd be stuck
+    // TODO: find/implement a channel that just throws away the oldest message when it is full, like a ring-buffer
+    source_is_connected: Arc<AtomicBool>,
+    serializer: EventSerializer,
+}
+
+impl AmiSink {
+    pub fn new(
+        response_tx: Sender<SourceReply>,
+        reply_tx: ReplySender,
+        config: Config,
+        source_is_connected: Arc<AtomicBool>,
+    ) -> Self {
+        let concurrency_cap = ConcurrencyCap::new(config.concurrency, reply_tx.clone());
+
+        Self {
+            request_counter: 1, // always start by 1, 0 is DEFAULT_STREAM_ID and this might interfere with custom codecs
+            response_tx,
+            reply_tx,
+            client: None,
+            config,
+            origin_uri: EventOriginUri {
+                scheme: String::from("azure_monitor_ingest_writer"),
+                host: String::from("dummy"), // will be replaced in `on_event`
+                port: None,
+                path: vec![],
+            },
+            concurrency_cap,
+            source_is_connected,
+            serializer: json_serializer(),
+        }
+    }
+}
+
+#[allow(clippy::unwrap_used)] // NOTE we know that the codec is always present/correct
```

Comment:
let's still not do that, it introduces headache later on if we ever extend the function. We an bubble up the error